File: backend/app/services/views/workforce.py
"""Workforce view — vertical content seam (Module 18a).

The one place caregiver *compliance and capacity* meaning lives on the server: what
"expiring" means, how much a caregiver could work vs. how much they are booked, and
which credentials need attention today. Core code never imports this — the seam
router (`routers/workforce.py`), the schedule router's roster, and the caregiver
tool handlers (`services/tools/entities.py`, also seam) are the only readers.

The Roster rides the Caregivers surface (it is NOT a fifth sanctioned vertical
surface): this file is a re-templating-seam member alongside `views/caregivers.py`,
`views/matching.py`, `views/clients.py`, and the entity migration.

Three things live here that a different vertical would rewrite wholesale:

  * CREDENTIAL STATUS — `credential_status()`, computed at READ time from
    `expires_at` and the in-seam `EXPIRING_DAYS` constant (the M16 `evv_flag`
    precedent, user-locked). No stored status column, no detector loop, no LLM: a
    credential's standing is a function of its date and today, and deriving it on
    read means it can never go stale after a renewal is entered.

  * UTILIZATION MATH — `available_week_hours()` / `roster_rows()` /
    `roster_metrics()`. Deterministic SQL and arithmetic, no LLM near the numbers
    (CLAUDE.md). Available hours come from the same `resources.availability` jsonb
    the matcher reads (via `matching._hm`), so "available" means one thing across
    the board, the matcher, and this page. A caregiver who has declared NO
    availability gets `utilization = None`, never 0% — we don't know their
    capacity, and printing 0% would read as "idle" when it means "unknown".

  * THE ACTIVE/INACTIVE LINE — an inactive caregiver is excluded from matching
    candidates (`views/matching.rank_candidates`) and from the schedule board's
    roster (`routers/schedule._roster`). `roster_rows()` here is the ONE surface
    that lists everyone, because the Roster tab is where you go to bring someone
    back. Compliance counts in `roster_metrics()` deliberately ignore inactive
    caregivers' credentials — a lapsed CPR for someone who isn't working is not a
    compliance emergency.
"""
from __future__ import annotations

from datetime import date, datetime, timedelta

from psycopg.rows import dict_row

from .matching import _hm, week_hours_map

# How far ahead a credential counts as "expiring". In-seam constant, not config:
# one client, and the number is a policy statement the office user should be able
# to read in the code (the EVV grace-window precedent).
EXPIRING_DAYS = 60

# Resource lifecycle. Not a delete — home-care staff leave and come back, and their
# visit history has to survive either way.
RESOURCE_STATUSES: list[dict] = [
    {"key": "active", "label": "Active"},
    {"key": "inactive", "label": "Inactive"},
]
RESOURCE_STATUS_KEYS: list[str] = [s["key"] for s in RESOURCE_STATUSES]
_RESOURCE_STATUS_LABELS: dict[str, str] = {s["key"]: s["label"] for s in RESOURCE_STATUSES}

# Credential states, in the order the UI stacks them (worst first).
CREDENTIAL_STATUSES: list[str] = ["expired", "expiring", "valid", "no_expiry"]


def is_valid_resource_status(status: str | None) -> bool:
    return status in _RESOURCE_STATUS_LABELS


def resource_status_label(status: str | None) -> str:
    """Plain label for a resource status. Falls back to the raw value so an
    unrecognized status never crashes a summary or a task title."""
    if status is None:
        return "—"
    return _RESOURCE_STATUS_LABELS.get(status, status)


# ---------------------------------------------------------------------------
# credential status — the read-time derivation
# ---------------------------------------------------------------------------
def credential_status(expires_at: date | None, today: date | None = None) -> str:
    """'no_expiry' | 'expired' | 'expiring' | 'valid' for one credential.

    Pure function of the expiry date and the day it is read on — nothing is stored.
    A null `expires_at` is a credential that does not renew (a one-time sign-off),
    which is a distinct state from "valid until a date", so the UI can say so.
    Today counts as expiring, not expired: it is valid through its last day.
    """
    if expires_at is None:
        return "no_expiry"
    ref = today or date.today()
    if isinstance(expires_at, datetime):
        expires_at = expires_at.date()
    if expires_at < ref:
        return "expired"
    if expires_at <= ref + timedelta(days=EXPIRING_DAYS):
        return "expiring"
    return "valid"


def days_until(expires_at: date | None, today: date | None = None) -> int | None:
    """Whole days until expiry (negative once past). None for a no-expiry row."""
    if expires_at is None:
        return None
    ref = today or date.today()
    if isinstance(expires_at, datetime):
        expires_at = expires_at.date()
    return (expires_at - ref).days


# ---------------------------------------------------------------------------
# capacity math
# ---------------------------------------------------------------------------
def available_week_hours(availability: dict | None) -> float | None:
    """Declared hours per week from the `resources.availability` jsonb
    ({"mon": ["08:00-16:00"], …}), summing every parseable window across every day.

    `None` — not 0.0 — when nothing is declared: an empty availability means we do
    not know this caregiver's capacity, and a 0 would make utilization look like a
    real number (and divide by zero). Malformed windows are skipped rather than
    raising; the matcher already treats the jsonb as best-effort.
    """
    if not availability:
        return None
    total = 0.0
    found = False
    for ranges in availability.values():
        for r in (ranges or []):
            parts = str(r).split("-")
            if len(parts) != 2:
                continue
            lo, hi = _hm(parts[0]), _hm(parts[1])
            if lo is None or hi is None or hi <= lo:
                continue
            total += (hi - lo) / 60.0
            found = True
    return round(total, 2) if found else None


def utilization(scheduled_hours: float, available_hours: float | None) -> float | None:
    """Scheduled ÷ available, as a percentage. `None` when capacity is unknown.
    Not capped — a caregiver booked past their declared availability is exactly
    what the Roster tab needs to show, so >100% is a real, displayable number."""
    if not available_hours:
        return None
    return round(100.0 * scheduled_hours / available_hours, 1)


# ---------------------------------------------------------------------------
# roster + metrics
# ---------------------------------------------------------------------------
# Every credential in the tenant, worst-dated first — one query for the whole
# roster (low tens of caregivers, so a per-row fetch would be pure overhead).
_CREDENTIALS_SQL = """
select rc.id, rc.resource_id, rc.qualification_id, q.name as qualification_name,
       rc.issued_at, rc.expires_at, rc.notes
  from public.resource_credentials rc
  join public.qualifications q on q.id = rc.qualification_id
 order by rc.expires_at nulls last, q.name
"""


def _credential_out(row: dict, today: date) -> dict:
    return {
        "id": str(row["id"]),
        "resource_id": str(row["resource_id"]),
        "qualification_id": str(row["qualification_id"]),
        "qualification_name": row["qualification_name"],
        "issued_at": row["issued_at"],
        "expires_at": row["expires_at"],
        "status": credential_status(row["expires_at"], today),
        "days_left": days_until(row["expires_at"], today),
        "notes": row["notes"],
    }


async def roster_rows(conn, ref: date | datetime | str | None = None) -> list[dict]:
    """Every caregiver — ACTIVE AND INACTIVE — with capacity and credentials.

    This is the one surface that lists inactive people: the Roster tab is where an
    office user goes to reactivate someone or check a departed caregiver's file.
    The board and the matcher both filter to active.

    `ref` picks the ISO week `hours_this_week` is measured over (defaults to today),
    reusing `matching.week_hours_map` so the number matches the board exactly.
    """
    ref = ref or date.today()
    hours = await week_hours_map(conn, ref)
    today = date.today()

    async with conn.cursor(row_factory=dict_row) as cur:
        await cur.execute(
            """select id, name, phone, email, status, address, zip, languages,
                      traits, qualification_ids, region_ids, availability
                 from public.resources order by name"""
        )
        resources = await cur.fetchall()
        await cur.execute(_CREDENTIALS_SQL)
        cred_rows = await cur.fetchall()

    by_resource: dict[str, list[dict]] = {}
    for row in cred_rows:
        by_resource.setdefault(str(row["resource_id"]), []).append(
            _credential_out(row, today)
        )

    out: list[dict] = []
    for r in resources:
        rid = str(r["id"])
        scheduled = round(hours.get(rid, 0.0), 2)
        available = available_week_hours(r["availability"])
        out.append({
            "id": rid,
            "name": r["name"],
            "phone": r["phone"],
            "email": r["email"],
            "status": r["status"],
            # address/zip ride along so the shared CaregiverDrawer can be opened
            # from the Roster tab without a second fetch — and so saving from
            # there can't blank a field the drawer never received.
            "address": r["address"],
            "zip": r["zip"],
            "languages": list(r["languages"] or []),
            "traits": list(r["traits"] or []),
            "qualification_ids": [str(x) for x in (r["qualification_ids"] or [])],
            "region_ids": [str(x) for x in (r["region_ids"] or [])],
            "availability": r["availability"] or {},
            "hours_this_week": scheduled,
            "available_hours": available,
            "utilization": utilization(scheduled, available),
            "credentials": by_resource.get(rid, []),
        })
    return out


async def roster_metrics(
    conn, ref: date | datetime | str | None = None, rows: list[dict] | None = None
) -> dict:
    """Headline compliance + capacity numbers for the Roster tab's strip.

    Computed from `roster_rows` (one pass, one definition of every number the page
    shows). `avg_utilization` averages only ACTIVE caregivers who have declared
    availability — including unknowns as zeroes would drag the number toward a
    meaningless floor. Credential counts likewise cover active caregivers only.

    Pass `rows` when the caller already has them (the router serves both in one
    response) so the roster query runs once, not twice.
    """
    if rows is None:
        rows = await roster_rows(conn, ref)
    active = [r for r in rows if r["status"] == "active"]

    utils = [r["utilization"] for r in active if r["utilization"] is not None]
    creds = [c for r in active for c in r["credentials"]]

    return {
        "active_count": len(active),
        "inactive_count": len(rows) - len(active),
        "avg_utilization": round(sum(utils) / len(utils), 1) if utils else None,
        "expiring_count": sum(1 for c in creds if c["status"] == "expiring"),
        "expired_count": sum(1 for c in creds if c["status"] == "expired"),
        "credential_count": len(creds),
    }


async def expiring_credentials(conn, days_ahead: int = EXPIRING_DAYS) -> list[dict]:
    """Active caregivers' credentials expiring within `days_ahead` days, soonest
    first — ALREADY-EXPIRED rows included, because "your CPR lapsed last week" is
    the most urgent version of this question, not a separate one.

    Each row is plain-language ready (caregiver + credential names, no UUIDs in the
    text) so the safe `list_expiring_credentials` tool and the daily digest
    automation can render it without a second lookup.
    """
    days_ahead = max(1, min(int(days_ahead), 365))
    today = date.today()
    async with conn.cursor(row_factory=dict_row) as cur:
        await cur.execute(
            """select rc.id, rc.resource_id, r.name as caregiver, r.phone,
                      q.name as credential, rc.issued_at, rc.expires_at, rc.notes
                 from public.resource_credentials rc
                 join public.resources r on r.id = rc.resource_id
                 join public.qualifications q on q.id = rc.qualification_id
                where r.status = 'active'
                  and rc.expires_at is not null
                  and rc.expires_at <= current_date + %(days)s::int
                order by rc.expires_at, r.name""",
            {"days": days_ahead},
        )
        rows = await cur.fetchall()

    return [
        {
            "id": str(r["id"]),
            "resource_id": str(r["resource_id"]),
            "caregiver": r["caregiver"],
            "phone": r["phone"],
            "credential": r["credential"],
            "issued_at": r["issued_at"],
            "expires_at": r["expires_at"],
            "days_left": days_until(r["expires_at"], today),
            "status": credential_status(r["expires_at"], today),
            "notes": r["notes"],
        }
        for r in rows
    ]


def describe_expiry(row: dict) -> str:
    """One plain-language clause for a digest line: "Maria Santos's CPR expires in
    30 days" / "…expired 10 days ago" / "…expires today"."""
    days = row.get("days_left")
    who = f"{row['caregiver']}'s {row['credential']}"
    if days is None:
        return f"{who} has no expiry on file"
    if days < 0:
        n = abs(days)
        return f"{who} expired {n} day{'s' if n != 1 else ''} ago"
    if days == 0:
        return f"{who} expires today"
    return f"{who} expires in {days} day{'s' if days != 1 else ''}"

File: frontend/src/lib/workforce.ts
// Workforce & compliance — vertical content seam (Module 18), the frontend mirror
// of backend/app/services/views/workforce.py. Credential/status labels and tones,
// utilization and days-left formatting, and the roster sort all live here (not in
// core UI) so the workforce surface stays re-templatable, alongside lib/leads.ts /
// lib/clients.ts / lib/caregivers.ts / lib/referrals.ts. Pure + vitest-covered.
//
// This module FORMATS server-computed values. It never derives a credential
// status or a utilization percentage — those come from the seam, so the badge on
// screen and the number in a digest task can never disagree.
import type {
  Credential,
  CredentialStatus,
  ResourceStatus,
  RosterCaregiver,
} from "@/lib/api";

// Badge variants this config uses (subset of the ui/badge variants).
type BadgeTone = "success" | "warning" | "destructive" | "secondary";

export interface CredentialStatusMeta {
  label: string;
  badge: BadgeTone;
  dot: string; // dot color class for chips and ui/Select options
  rank: number; // sort weight — worst first, so problems surface at a glance
}

// expired = destructive (they may legally not be able to work a shift),
// expiring = warning (act now), valid = success, no_expiry = muted (a one-time
// sign-off is fine, but it is NOT the same as "valid until a date").
export const CREDENTIAL_STATUS_META: Record<CredentialStatus, CredentialStatusMeta> = {
  expired: { label: "Expired", badge: "destructive", dot: "bg-destructive", rank: 0 },
  expiring: { label: "Expiring", badge: "warning", dot: "bg-warning", rank: 1 },
  valid: { label: "Valid", badge: "success", dot: "bg-success", rank: 2 },
  no_expiry: { label: "No expiry", badge: "secondary", dot: "bg-muted-foreground", rank: 3 },
};

export const CREDENTIAL_STATUSES: CredentialStatus[] = [
  "expired",
  "expiring",
  "valid",
  "no_expiry",
];

const UNKNOWN_CREDENTIAL_META: CredentialStatusMeta = {
  label: "Unknown",
  badge: "secondary",
  dot: "bg-muted-foreground",
  rank: 4,
};

export function credentialMeta(status: string | null | undefined): CredentialStatusMeta {
  if (!status) return UNKNOWN_CREDENTIAL_META;
  return (
    CREDENTIAL_STATUS_META[status as CredentialStatus] ?? {
      ...UNKNOWN_CREDENTIAL_META,
      label: status,
    }
  );
}

export interface ResourceStatusMeta {
  label: string;
  badge: BadgeTone;
  dot: string;
}

// Inactive is muted, never destructive — leaving is a normal lifecycle event, not
// an error state, and the person's history is kept either way.
export const RESOURCE_STATUS_META: Record<ResourceStatus, ResourceStatusMeta> = {
  active: { label: "Active", badge: "success", dot: "bg-success" },
  inactive: { label: "Inactive", badge: "secondary", dot: "bg-muted-foreground" },
};

export const RESOURCE_STATUSES: ResourceStatus[] = ["active", "inactive"];

export function resourceStatusMeta(status: string | null | undefined): ResourceStatusMeta {
  if (!status) return RESOURCE_STATUS_META.active;
  return (
    RESOURCE_STATUS_META[status as ResourceStatus] ?? {
      label: status,
      badge: "secondary",
      dot: "bg-muted-foreground",
    }
  );
}

// --- Formatting --------------------------------------------------------------
function trimNumber(n: number): string {
  const rounded = Math.round(n * 10) / 10;
  return Number.isInteger(rounded) ? String(rounded) : rounded.toFixed(1);
}

// Utilization is null when capacity is unknown (no declared availability) — that
// is NOT 0%, so it renders as an em dash. Display caps at 999% so one absurd
// number can't blow out the column; the underlying value is never clamped.
const UTILIZATION_DISPLAY_CAP = 999;

export function fmtUtilization(pct: number | null | undefined): string {
  if (pct == null) return "—";
  if (pct > UTILIZATION_DISPLAY_CAP) return `${UTILIZATION_DISPLAY_CAP}%+`;
  return `${trimNumber(pct)}%`;
}

// Bar width for the utilization meter, 0–100. Over-booked caregivers peg the bar
// full; the number beside it carries the real magnitude.
export function utilizationBarPct(pct: number | null | undefined): number {
  if (pct == null || pct <= 0) return 0;
  return Math.min(100, pct);
}

// Over their declared availability — the bar and the number both go warning-toned.
export function isOverbooked(pct: number | null | undefined): boolean {
  return pct != null && pct > 100;
}

export function fmtHours(n: number | null | undefined): string {
  if (n == null) return "—";
  return `${trimNumber(n)}h`;
}

// "expires in 30 d" / "expired 10 d ago" / "expires today" / "no expiry".
// Mirrors the backend's describe_expiry phrasing, abbreviated for a table cell.
export function fmtDaysLeft(days: number | null | undefined): string {
  if (days == null) return "no expiry";
  if (days < 0) {
    const n = Math.abs(days);
    return `expired ${n} d ago`;
  }
  if (days === 0) return "expires today";
  return `expires in ${days} d`;
}

// "Mar 4, 2027" — a date the office user reads, from the API's YYYY-MM-DD string.
// Parsed as a LOCAL date (not `new Date("2027-03-04")`, which is UTC midnight and
// renders as the previous day west of Greenwich).
export function fmtDate(iso: string | null | undefined): string {
  if (!iso) return "—";
  const [y, m, d] = iso.split("-").map(Number);
  if (!y || !m || !d) return iso;
  return new Date(y, m - 1, d).toLocaleDateString(undefined, {
    year: "numeric",
    month: "short",
    day: "numeric",
  });
}

// --- Sorting -----------------------------------------------------------------
// Credentials worst-first, then soonest expiry, then name — so a caregiver's
// problem credential is always the first chip in their row.
export function sortCredentials(credentials: Credential[]): Credential[] {
  return [...credentials].sort((a, b) => {
    const rank = credentialMeta(a.status).rank - credentialMeta(b.status).rank;
    if (rank !== 0) return rank;
    const ad = a.days_left ?? Number.MAX_SAFE_INTEGER;
    const bd = b.days_left ?? Number.MAX_SAFE_INTEGER;
    if (ad !== bd) return ad - bd;
    return a.qualification_name.localeCompare(b.qualification_name);
  });
}

// Roster order: active caregivers first (they're who you staff), then by name.
export function sortRoster(rows: RosterCaregiver[]): RosterCaregiver[] {
  return [...rows].sort((a, b) => {
    if (a.status !== b.status) return a.status === "active" ? -1 : 1;
    return a.name.localeCompare(b.name);
  });
}

// Client-side search + status filter (roster scale is low tens — no server round
// trip). Matches name, phone, email, and credential names, case-insensitively.
export function filterRoster(
  rows: RosterCaregiver[],
  { search, status }: { search?: string; status?: string },
): RosterCaregiver[] {
  const q = (search ?? "").trim().toLowerCase();
  return rows.filter((r) => {
    if (status && r.status !== status) return false;
    if (!q) return true;
    const haystack = [
      r.name,
      r.phone ?? "",
      r.email ?? "",
      ...r.credentials.map((c) => c.qualification_name),
    ]
      .join(" ")
      .toLowerCase();
    return haystack.includes(q);
  });
}
